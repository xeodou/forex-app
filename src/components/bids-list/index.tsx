import { CurrencyRate } from "lib/forex-api/client";
import { FC } from "react";

export const BidsList: FC<{ rates: CurrencyRate[] }> = ({ rates }) => {
  return (
    <div className="bids-list w-full">
      <header className="flex justify-around uppercase">
        <span>CUR</span>
        <span>Time</span>
        <span>Ask</span>
        <span>Bid</span>
      </header>
      <div className="h-60 overflow-y-auto flex flex-col">
        {rates.map((rate, index) => (
          <ul
            key={rate.time_stamp}
            className={`flex justify-around ${index % 2 === 0 ? "bg-slate-300" : "bg-slate-100 "}`}
          >
            <li>{`${rate.from}/${rate.to}`}</li>
            <li>{new Date(rate.time_stamp).toLocaleString()}</li>
            <li>{rate.ask}</li>
            <li>{rate.bid}</li>
          </ul>
        ))}
      </div>
    </div>
  );
};
