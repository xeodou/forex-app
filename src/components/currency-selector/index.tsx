import { FC } from "react";
import { currencies } from "utils/currencies";

export interface CurrencyHeaderProps {
  // On selected currency changes
  onSelected?: (currency: string) => void;
  // Specific the default selected currency
  value?: string;
}

export const CurrencySelector: FC<CurrencyHeaderProps> = ({ onSelected, value }) => {
  const onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onSelected && onSelected(event.target.value);
  };

  return (
    <select
      onChange={onChange}
      value={value}
      className="h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm "
    >
      {currencies.map((currency) => (
        <option key={currency} value={currency}>
          {currency}
        </option>
      ))}
    </select>
  );
};
