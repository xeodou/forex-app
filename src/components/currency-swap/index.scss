.currency-header {
  &--indicator {
    width: 30px;
  }
}
