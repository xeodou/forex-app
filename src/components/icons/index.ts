export * from "./icon-swap";
export * from "./icon-refresh";
export * from './icon-close';
