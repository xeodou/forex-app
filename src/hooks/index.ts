export * from "./use-currency-pair";
export * from "./use-forex-rates";
