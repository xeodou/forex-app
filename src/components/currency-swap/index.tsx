import { CurrencySelector } from "components/currency-selector";
import { IconSwap } from "components/icons";
import { useCurrencyPair } from "hooks/use-currency-pair";

export const CurrencySwap = ({}) => {
  const [currencyPair, setCurrencyPair] = useCurrencyPair();

  const onFromCurrencyChange = (from: string) => {
    setCurrencyPair([from, currencyPair[1]]);
  };

  const onToCurrencyChange = (to: string) => {
    setCurrencyPair([currencyPair[0], to]);
  };

  const onSwitchClick = () => {
    setCurrencyPair([currencyPair[1], currencyPair[0]]);
  };

  return (
    <div className="flex items-center">
      <CurrencySelector onSelected={onFromCurrencyChange} value={currencyPair[0]} />
      <button
        type="button"
        aria-label="Switch currencies"
        className="font-medium text-sm p-2.5 text-center inline-flex items-center"
        onClick={onSwitchClick}
      >
        <IconSwap></IconSwap>
      </button>
      <CurrencySelector onSelected={onToCurrencyChange} value={currencyPair[1]} />
    </div>
  );
};
