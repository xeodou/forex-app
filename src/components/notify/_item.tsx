import { IconClose } from "components/icons"
import { NotifyItem, NotifyType } from "hooks"
import { FC } from "react"

export interface ItemProps {
  item: NotifyItem;
  onDismiss: (id: number) => void;
}

export const Item: FC<ItemProps> = ({item, onDismiss}) => {
  return (
    <div className={`rounded-md p-4 ${item.type === NotifyType.Error ? 'bg-yellow-50' : 'bg-green-50'}`}>
      <div className="flex">
        <div className="ml-3">
          <p className={`text-sm font-medium ${item.type === NotifyType.Error ? 'text-yellow-800' : 'text-green-800'}`}>
            {item.message}
          </p>
        </div>
        <div className="ml-auto pl-3">
              <div className="-mx-1.5 -my-1.5">
                <button type="button" className="inline-flex align-baseline" onClick={() => onDismiss(item.id as number)}>
                  <span className="sr-only">Dismiss</span>
                  <IconClose aria-hidden="true" />
                </button>
              </div>
            </div>
      </div>
    </div>
  )
}
