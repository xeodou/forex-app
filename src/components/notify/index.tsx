import { useNotify } from "hooks";
import { Item } from "./_item";

export const Notify = () => {
  const { stack, removeNotifyItem } = useNotify();

  const lastItem = stack[stack.length - 1];

  const onDismiss = (id: number) => {
    removeNotifyItem(id);
  };

  return <div className="notify-container">{lastItem && <Item item={lastItem} onDismiss={onDismiss} />}</div>;
};
