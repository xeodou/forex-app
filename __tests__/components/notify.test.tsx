import { act, fireEvent } from "@testing-library/react";
import { render, unmountComponentAtNode } from "react-dom";
import pretty from "pretty";
import { NotifyContextProvider, NotifyItem, NotifyType } from "hooks";
import { Notify } from "components/notify";
import { FC } from "react";

describe("notify", () => {
  let container: HTMLDivElement;

  const Wrapper: FC<{initialValue: NotifyItem[]}> = ({ initialValue }) => (
    <NotifyContextProvider initialValue={initialValue}>
      <Notify />
    </NotifyContextProvider>
  );

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null as unknown as HTMLDivElement;
  });

  it("should render the notify component", () => {
    const item: NotifyItem = {
      id: Date.now(),
      type: NotifyType.Success,
      message: 'lorem'
    }
    act(() => {
      render(<Wrapper initialValue={[item]}/>, container);
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();

    expect(container.querySelector("div.notify-container")).toBeTruthy();
    expect(container.querySelector('p')?.textContent).toEqual('lorem');

    const button = container.querySelector("button") as HTMLButtonElement;
    expect(button).toBeTruthy();

    fireEvent.click(button);

    expect(container.querySelector('p')).toBeFalsy();
  });
});
