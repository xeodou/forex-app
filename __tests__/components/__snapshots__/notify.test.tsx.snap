// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notify should render the notify component 1`] = `
"<div class=\\"notify-container\\">
  <div class=\\"rounded-md p-4 bg-green-50\\">
    <div class=\\"flex\\">
      <div class=\\"ml-3\\">
        <p class=\\"text-sm font-medium text-green-800\\">lorem</p>
      </div>
      <div class=\\"ml-auto pl-3\\">
        <div class=\\"-mx-1.5 -my-1.5\\"><button type=\\"button\\" class=\\"inline-flex align-baseline\\"><span class=\\"sr-only\\">Dismiss</span><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"h-6 w-6\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" stroke-width=\\"2\\">
              <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M6 18L18 6M6 6l12 12\\"></path>
            </svg></button></div>
      </div>
    </div>
  </div>
</div>"
`;
